File: src/repositories/authorizationRepository.js
'use strict';

const { randomUUID } = require('crypto');

/**
 * In-memory authorisation store. The demo service keeps no external database.
 */
const store = new Map();
const failures = {};

const authorizationRepository = {
  save(record) {
    const saved = {
      authorizationId: randomUUID(),
      authorizedAt: new Date().toISOString(),
      ...record,
    };
    store.set(saved.authorizationId, saved);
    return saved;
  },

  findByIdempotencyKey(idempotencyKey) {
    if (!idempotencyKey) return null;
    return [...store.values()].find((r) => r.idempotencyKey === idempotencyKey) || null;
  },

  findById(authorizationId) {
    return store.get(authorizationId) || null;
  },

  findByInvoice(invoiceId) {
    return [...store.values()].filter((record) => record.invoiceId === invoiceId);
  },

  countByNetwork() {
    return [...store.values()].reduce((counts, record) => {
      counts[record.network] = (counts[record.network] || 0) + 1;
      return counts;
    }, {});
  },

  recordFailure(network) {
    failures[network] = (failures[network] || 0) + 1;
  },

  failuresByNetwork() {
    return { ...failures };
  },

  size() {
    return store.size;
  },
};

module.exports = { authorizationRepository };

File: src/main/kotlin/com/beaconstone/payments/repositories/AuthorizationRepository.kt
package com.beaconstone.payments.repositories

import java.time.Instant
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

data class AuthorizationRecord(
    val authorizationId: String = UUID.randomUUID().toString(),
    val authorizedAt: String = Instant.now().toString(),
    val paymentId: String? = null,
    val invoiceId: String? = null,
    val network: String? = null,
    val profileId: String? = null,
    val captureMode: String? = null,
    val installmentPlan: String? = null,
    val authorizationCode: String? = null,
    val amountMinor: Int = 0,
    val capturedMinor: Int = 0,
    val currency: String = "USD",
    val status: String = "authorized",
    val idempotencyKey: String? = null,
)

/**
 * In-memory authorisation store. The demo service keeps no external database.
 */
object authorizationRepository {
    private val store = ConcurrentHashMap<String, AuthorizationRecord>()
    private val failures = ConcurrentHashMap<String, Int>()

    fun save(record: AuthorizationRecord): AuthorizationRecord {
        store[record.authorizationId] = record
        return record
    }

    fun findById(authorizationId: String): AuthorizationRecord? = store[authorizationId]

    fun findByIdempotencyKey(idempotencyKey: String?): AuthorizationRecord? {
        if (idempotencyKey.isNullOrBlank()) return null
        return store.values.find { it.idempotencyKey == idempotencyKey }
    }

    fun findByInvoice(invoiceId: String): List<AuthorizationRecord> =
        store.values.filter { it.invoiceId == invoiceId }

    fun recordFailure(network: String?) {
        val key = network ?: "unknown"
        failures.merge(key, 1, Int::plus)
    }

    fun failuresByNetwork(): Map<String, Int> = failures.toMap()

    fun countByNetwork(): Map<String, Int> =
        store.values.groupingBy { it.network ?: "unknown" }.eachCount()

    fun size(): Int = store.size
}
